### PR TITLE
Add a message parser

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Verifying message hashes isn't too meaningful if the user suspects the host computer may be compromised.

#### Proposed changes

Work towards analyzing message contents from the Ledger Wallet. This PR includes a recursive descent parser for Solana messages. It's not hooked up to anything yet, but enough code on its own for a review.

### CODE REVIEW

File: src/parseMessage.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#ifndef PARSE_MESSAGE_H
```
Comment (jackcmay) [inline src/parseMessage.h]: How about `#pragma once`?

File: src/parseMessageTest.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+   assert(parse_blockhash(&parser, &blockhash) == 0);
+   assert(parser.buffer_length == 0);
+   assert(parser.buffer == message + 32);
+}
+
+void test_parse_blockhash_too_short() {
+   uint8_t message[31]; // <--- Too short!
+   MessageParser parser = {message, sizeof(message)};
+   Blockhash* blockhash;
+   assert(parse_blockhash(&parser, &blockhash) == 1);
```
Comment (jackcmay) [inline src/parseMessageTest.c]: Test that the blockhash contains the correct data?

File: src/parseMessageTest.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+void test_parse_blockhash() {
+   uint8_t message[32];
+   MessageParser parser = {message, sizeof(message)};
+   Blockhash* blockhash;
+   assert(parse_blockhash(&parser, &blockhash) == 0);
+   assert(parser.buffer_length == 0);
+   assert(parser.buffer == message + 32);
+}
```
Comment (jackcmay) [inline src/parseMessageTest.c]: Test that the blockhash contains the correct data?

File: src/parseMessageTest.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+void test_parse_pubkeys() {
+   uint8_t message[33] = {1};
+   MessageParser parser = {message, sizeof(message)};
+   Pubkey* pubkeys;
+   size_t pubkeys_length;
+   assert(parse_pubkeys(&parser, &pubkeys, &pubkeys_length) == 0);
+   assert(parser.buffer_length == 0);
+   assert(parser.buffer == message + 33);
+}
```
Comment (jackcmay) [inline src/parseMessageTest.c]: Test that the pubkey contains the correct data?

File: src/parseMessageTest.c [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+   Blockhash* blockhash;
+   assert(parse_blockhash(&parser, &blockhash) == 1);
+}
+
+void test_parse_data() {
+   uint8_t message[] = {1, 2};
+   MessageParser parser = {message, sizeof(message)};
+   uint8_t* data;
+   size_t data_length;
+   assert(parse_data(&parser, &data, &data_length) == 0);
```
Comment (jackcmay) [inline src/parseMessageTest.c]: Maybe I'm missing something, isn't `parse_data()` a private function of `parseMesssage.c`?

File: src/parseMessage.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#define PARSE_MESSAGE_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+typedef struct MessageParser {
```
Comment (jackcmay) [inline src/parseMessage.h]: nit: This isn't really the parser, could call it `MessageBytes` or `RawMessage`, `MessageStream`, or ...

File: src/parseMessage.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+int parse_length(MessageParser* parser, size_t* value);
+
+int parse_message_header(MessageParser* parser, MessageHeader* header);
+
+int parse_pubkeys(MessageParser* parser, Pubkey** pubkeys, size_t* pubkeys_length);
+
+int parse_blockhash(MessageParser* parser, Blockhash** blockhash);
+
+int parse_instruction(MessageParser* parser, Instruction* instruction);
+
```
Comment (jackcmay) [inline src/parseMessage.h]: Could combine the above into a `parse_message()`.  I guess that could result in an unwieldy structure...

File: src/parseMessageTest.c (hunk 5 above)
Comment (garious) [reply src/parseMessageTest.c]: Followed an old BREW convention here, which was to have the unit tests `#include` the source file, such that one can test private functions.

Comment (jackcmay) [reply src/parseMessageTest.c]: Ah, yeah, I like that

File: src/parseMessage.h (hunk 6 above)
Comment (garious) [reply src/parseMessage.h]: I'll give it some thought.  It seems more that "Message" is the wrong word than "Parser", no?

Comment (jackcmay) [reply src/parseMessage.h]: Maybe both, that structure's name contains a verb though it doesn't perform any action.  Just a nit, take it or leave it :-)

File: src/parseMessage.h (hunk 7 above)
Comment (garious) [reply src/parseMessage.h]: Right, that's exactly the issue.  By stopping at `parse_instruction()`, it becomes the job of the caller to iterate over the instructions one by one and extract the information they need.  I'm trying to keep this as a zero-copy, zero-alloc parser.  It's not obvious to me what the function signature of `parse_message` would be.  What's the thinking on VLAs these days?

Comment (jackcmay) [reply src/parseMessage.h]: You can probably do something similar to what we do currently with `SolKeyedAccount` for C BPF programs.  `Message` would be more complicated but I think the same concepts apply.  A collection of pointers, lengths, and non-pointer members.

```c
/**
 * Keyed Account
 */
typedef struct {
  SolPubkey *key;        /** Public key of the account */
  bool is_signer;        /** Transaction was signed by this account's key? */
  bool is_writable;      /** Is the account writable? */
  uint64_t *lamports;    /** Number of lamports owned by this account */
  uint64_t userdata_len; /** Length of data in bytes */
  uint8_t *userdata;     /** On-chain data within this account */
  SolPubkey *owner;      /** Program that owns this account */
} SolKeyedAccount;
```

Where for example `SolPubkey` is:

```c
/**
 * Size of Public key in bytes
 */
#define SIZE_PUBKEY 32

/**
 * Public key
 */
typedef struct {
  uint8_t x[SIZE_PUBKEY];
} SolPubkey;
```

Comment (garious) [reply src/parseMessage.h]: I see, thanks.  P.S. "userdata" !?

File: src/parseMessageTest.c (hunk 4 above)
Comment (garious) [reply src/parseMessageTest.c]: Fixed

File: src/parseMessageTest.c (hunk 3 above)
Comment (garious) [reply src/parseMessageTest.c]: Fixed

File: src/parseMessage.h (hunk 6 above)
Comment (garious) [reply src/parseMessage.h]: went with Parser

File: src/parseMessage.h (hunk 7 above)
Comment (jackcmay) [reply src/parseMessage.h]: Heh, yeah, needs updating, will do :-)

**Pull request merged**